torchtitan/parallelisms/parallelize_llama.py

```diff
@@ -295,142 +295,191 @@ def pipeline_llama_tracer(
 
-    # apply AC + torch.compile
+def apply_ac(model, world_mesh, parallel_dims, job_config: JobConfig):
+    """
+    Apply activation checkpointing to the model.
+    """
+
     ac_config = job_config.activation_checkpoint
-    enable_compile = job_config.training.compile
+    assert ac_config.mode in (
+        "full",
+        "selective",
+    ), f"{ac_config.mode=}, expecting full or selective"
+
     for layer_id, transformer_block in model.layers.named_children():
-        if ac_config.mode in ("full", "selective"):
-            transformer_block = checkpoint_wrapper(transformer_block, ac_config)
-        if enable_compile:
-            # turn on per-transformer block compile after AC wrapping and before FSDP
-            # TODO: dynamic shape have some issues so we turn it off for now.
-            # TODO: inline inbuilt nn modules does not work yet, enable it to accelarate
-            # compile time.
-            # torch._dynamo.config.inline_inbuilt_nn_modules = True
-            transformer_block = torch.compile(transformer_block, dynamic=False)
+        transformer_block = checkpoint_wrapper(transformer_block, ac_config)
         model.layers.register_module(layer_id, transformer_block)
 
-    if ac_config.mode in ("full", "selective"):
-        logger.info(f"Applied {ac_config.mode} activation checkpointing to the model")
-        if (
-            enable_compile
-            and ac_config.mode == "selective"
-            and ac_config.selective_ac_option == "op"
-        ):
-            # some temp flags for torch.compile enablement + SAC
-            torch._dynamo.config._experimental_support_context_fn_in_torch_utils_checkpoint = (
-                True
-            )
-    if enable_compile:
-        if job_config.model.norm_type == "fused_rmsnorm":
-            raise NotImplementedError(
-                "fused_rmsnorm not yet compatible with torch.compile. Please use layernorm or rmsnorm."
-            )
-        logger.info("Compiled each TransformerBlock with torch.compile")
+    logger.info(f"Applied {ac_config.mode} activation checkpointing to the model")
+    return model
 
-    # apply DP (FSDP2)
-    if parallel_dims.dp_enabled:
-        dp_mesh = world_mesh["dp"] if world_mesh.ndim > 1 else world_mesh
-        assert dp_mesh.mesh_dim_names == ("dp",), dp_mesh.mesh_dim_names
-        mp_policy = MixedPrecisionPolicy(
-            param_dtype=TORCH_DTYPE_MAP[job_config.training.mixed_precision_param],
-            reduce_dtype=TORCH_DTYPE_MAP[job_config.training.mixed_precision_reduce],
+
+def apply_compile(model, world_mesh, parallel_dims, job_config: JobConfig):
+    """
+    Apply torch.compile to the model.
+    """
+
+    assert job_config.training.compile, "torch.compile is not enabled"
+    if job_config.model.norm_type == "fused_rmsnorm":
+        raise NotImplementedError(
+            "fused_rmsnorm not yet compatible with torch.compile. Please use layernorm or rmsnorm."
```

Comment:
IMO we should put this type of condition (not directly related to compile itself) outside in `parallelize_llama`. (And eventually we will replace this `fused_rmsnorm` by compiled rmsnorm.)

Reply:
Hmm, same tradeoff bc it would be too detailed if put in `parallelize_llama`. Will wait for the eventual solution then.